### PR TITLE
Use expiration policy

### PR DESCRIPTION
**Pull request by AurelienFT**
## Linked Issues/PRs
Spec PR : https://github.com/FuelLabs/fuel-specs/pull/616

VM PR : https://github.com/FuelLabs/fuel-vm/pull/871

## Description
This PR use the new policy `expiration` in transaction pool to prune transactions when they are expired to prevent them to be included in a block.

Also verification is added when executing transaction in the executor.

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests
- [x] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [x] I have reviewed the code myself
- [x] I have created follow-up issues caused by this PR and linked them here

### After merging, notify other teams

[Add or remove entries as needed]

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)
- [ ] [Sway compiler](https://github.com/FuelLabs/sway/)
- [ ] [Platform documentation](https://github.com/FuelLabs/devrel-requests/issues/new?assignees=&labels=new+request&projects=&template=NEW-REQUEST.yml&title=%5BRequest%5D%3A+) (for out-of-organization contributors, the person merging the PR will do this)
- [ ] Someone else?

### CODE REVIEW

File: crates/services/txpool_v2/src/service.rs [hunk 1]
```diff
@@ -323,6 +324,29 @@ where
...
+                .range(..=new_height)
+                .map(|(k, _)| *k)
+                .collect::<Vec<_>>();
+            for height in range_to_remove {
+                let expired_txs = height_expiration_txs.remove(&height);
+                if let Some(expired_txs) = expired_txs {
+                    let mut tx_pool = self.pool.write();
+                    removed_txs
+                        .extend(tx_pool.remove_transaction_and_dependents(expired_txs));
+                }
```
Comment (rymnc) [inline crates/services/txpool_v2/src/service.rs:342]: is it possible that when you call `tx_pool.remove_transaction_and_dependents` you end up removing the same tx twice? i.e once as a dependent for another transaction and once on its own?

File: crates/services/txpool_v2/src/service.rs [hunk 2]
```diff
@@ -323,6 +324,29 @@ where
...
+            }
+        }
+        for tx in removed_txs {
```
Comment (rymnc) [inline crates/services/txpool_v2/src/service.rs:345]: would be nice to know if `removed_txs` are unique

File: crates/services/txpool_v2/src/service.rs [hunk 3]
```diff
@@ -764,6 +796,7 @@ where
     let pruner = TransactionPruner {
         txs_ttl: config.max_txs_ttl,
         time_txs_submitted: Arc::new(RwLock::new(VecDeque::new())),
+        height_expiration_txs: Arc::new(RwLock::new(BTreeMap::new())),
```
Comment (rymnc) [inline crates/services/txpool_v2/src/service.rs:799]: can't we use dashmap here?
will prevent usage of `RwLock`

also may be worth to think about usage of channels to send transactions that need to be pruned and then the `TransactionPruner` can handle them appropriately according to some configuration

File: crates/services/txpool_v2/src/tests/tests_service.rs [hunk 4]
```diff
@@ -236,6 +252,89 @@ async fn test_prune_transactions_the_oldest() {
     service.stop_and_await().await.unwrap();
 }
 
+#[tokio::test]
+async fn prune_expired_transactions() {
```
Comment (rymnc) [inline crates/services/txpool_v2/src/tests/tests_service.rs:256]: can you add a test that checks if the `tx_status_sender` does not send duplicate tx statuses for transactions that may have been removed twice acc to my earlier comment?

File: crates/services/txpool_v2/src/tests/universe.rs [hunk 5]
```diff
@@ -93,6 +94,7 @@ use super::mocks::{
 // the byte and gas price fees.
 pub const TEST_COIN_AMOUNT: u64 = 100_000_000u64;
 pub const GAS_LIMIT: Word = 100000;
+pub const DEFAULT_EXPIRATION_HEIGHT: BlockHeight = BlockHeight::new(1000);
```
Comment (rymnc) [inline crates/services/txpool_v2/src/tests/universe.rs:97]:
```suggestion
pub const DEFAULT_EXPIRATION_HEIGHT: BlockHeight = BlockHeight::new(10);
```
1_000 seems like state would change too much, no?

File: crates/client/src/client/schema/chain.rs [hunk 6]
```diff
@@ -472,6 +473,129 @@ impl TryFrom<GasCosts> for fuel_core_types::fuel_tx::GasCosts {
                 }
                 .into(),
             )),
+            GasCostsVersion::V2 => {
```
Comment (rymnc) [inline crates/client/src/client/schema/chain.rs:476]: curious why we need this change

**Review by rymnc — COMMENTED** [review]
left some comments

File: crates/services/txpool_v2/src/service.rs (hunk 1 above)
Comment (AurelienFT) [reply crates/services/txpool_v2/src/service.rs:342]: Yeah or even if a transaction is the dependency of two others. But in the implementation it will not remove twice because of https://github.com/FuelLabs/fuel-core/blob/ccae2d50c4d694a8c77500d2843a3322f694e80b/crates/services/txpool_v2/src/storage/graph.rs#L144

File: crates/services/txpool_v2/src/service.rs (hunk 2 above)
Comment (AurelienFT) [reply crates/services/txpool_v2/src/service.rs:345]: Yeah removed_txs should be unique in any cases because we don't want to push removal notification twice for the same id (it should'nt happens but who knows). I don't know if changing `removed_txs` to `Hashset` is ok because we iterate over it and might be performance consuming

File: crates/services/txpool_v2/src/tests/universe.rs (hunk 5 above)
Comment (AurelienFT) [reply crates/services/txpool_v2/src/tests/universe.rs:97]: This is a test onyl feature to be sure it doesn't change the existing tests that may be running for long. In the real world the default value is "no expiration"

File: crates/services/txpool_v2/src/service.rs (hunk 3 above)
Comment (AurelienFT) [reply crates/services/txpool_v2/src/service.rs:799]: Yeah we can. We didn't use it across the code yet but could be a great first place experiment but is it really useful to it if we want to change to channels etc (which could a great idea that I need to study). Are you ok if I just open an issue about improving the transaction pruner ?

File: crates/services/txpool_v2/src/service.rs (hunk 1 above)
Comment (rymnc) [reply crates/services/txpool_v2/src/service.rs:342]: okay :)

File: crates/services/txpool_v2/src/service.rs (hunk 3 above)
Comment (rymnc) [reply crates/services/txpool_v2/src/service.rs:799]: sure, feel free to do that :)

File: crates/services/txpool_v2/src/tests/universe.rs (hunk 5 above)
Comment (rymnc) [reply crates/services/txpool_v2/src/tests/universe.rs:97]: should the default value be "no expiration"? is that for backwards compatibility or some other reason?

**Review by rymnc — APPROVED** [review]

File: crates/services/txpool_v2/src/tests/universe.rs (hunk 5 above)
Comment (AurelienFT) [reply crates/services/txpool_v2/src/tests/universe.rs:97]: For tests I find it easier to write test in the future then and so they all shared a same constant instead of magical numbers in each tests that would endup in a constant

Comment (rymnc) [reply crates/services/txpool_v2/src/tests/universe.rs:97]: oh, I'm curious about the "real world" default for tx expiry, not the tests here :)

**Review by rymnc — APPROVED** [review]

File: crates/services/txpool_v2/src/tests/universe.rs (hunk 5 above)
Comment (AurelienFT) [reply crates/services/txpool_v2/src/tests/universe.rs:97]: In real world it's "no expiration"

**Review by Dentosal — APPROVED** [review]
